src/apps/polymer/IncentivizedPolymerEscrow.sol

```diff
@@ -0,0 +1,175 @@
+    // channelId: the universal channel id on the source chain, which can be used to identify the source chain.
+    function onRecvUniversalPacket(bytes32 channelId, UniversalPacket calldata packet)
+        external
+        onlyIbcMw
+        returns (AckPacket memory)
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes memory sourceImplementationIdentifier = abi.encodePacked(packet.srcPortAddr);
+
+        bytes memory receiveAck = _handleMessage(
+            bytes32(abi.encodePacked(packet.srcPortAddr)),
+            sourceImplementationIdentifier,
+            packet.appData,
+            feeRecipitent,
+            gasLimit
+        );
+
+        // Send ack:
+        return AckPacket({success: true, data: receiveAck});
+    }
+
+    // The escrow manages acks, so any message can be directly provided to _handleAck.
+    function onUniversalAcknowledgement(bytes32 channelId, UniversalPacket calldata packet, AckPacket calldata ack)
+        external
+        onlyIbcMw
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = ack.data;
+        bytes32 chainIdentifier = _getChannelId(channelId);
+        bytes memory destinationImplementationIdentifier = abi.encodePacked(packet.destPortAddr);
+
+        isVerifiedMessageHash[keccak256(rawMessage)] = VerifiedMessageHashContext({
+            chainIdentifier: chainIdentifier,
+            implementationIdentifier: destinationImplementationIdentifier
+        });
+        _handleAck(chainIdentifier, destinationImplementationIdentifier, rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // For timeouts, we need to construct the message.
+    function onTimeoutUniversalPacket(bytes32 channelId, UniversalPacket calldata packet) external onlyIbcMw {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = packet.appData;
+        _handleTimeout(_getDestChainId(channelId), rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // * Send to messaging_protocol
+    function _sendPacket(
+        bytes32 destinationChainIdentifier,
+        bytes memory, /* destinationImplementation */
+        bytes memory message
+    ) internal override returns (uint128 costOfsendPacketInNativeToken) {
+        // call `setChannelId` to set the channelId for the destination chain before calling this function.
+        bytes32 channelId = _getChannelId(destinationChainIdentifier);
+        // set timeoutTimestamp to 1 day from now. It's the dest chain's block time in nanoseconds since the epoch.
+        uint64 timeoutTimestamp = uint64(block.timestamp + 1 days) * 1e9;
+        IbcUniversalPacketSender(mw).sendUniversalPacket(
+            channelId, destinationChainIdentifier, message, timeoutTimestamp
```

Comment:
Does channelId encode the destination address?
If so, there may be a case where `setRemoteImplementation` should be overwritten by this implementation to initialize channel creation (if needed) when called. https://github.com/catalystdao/GeneralisedIncentives/blob/d9e39f02a0ea3d4ca8c4d680fff12b083afdd36e/src/IncentivizedMessageEscrow.sol#L92-L100

Reply:
Going off our discussion, the `_sendPacket` function cloud be implemented as:
```solidity
// * Send to messaging_protocol
  function _sendPacket(
      bytes32 destinationChainIdentifier,
      bytes memory destinationImplementation
      bytes memory message
  ) internal override returns (uint128 costOfsendPacketInNativeToken) {
      // call `setChannelId` to set the channelId for the destination chain before calling this function.
      bytes32 channelId = _getChannelId(destinationChainIdentifier);
      // set timeoutTimestamp to 1 day from now. It's the dest chain's block time in nanoseconds since the epoch.
      uint64 timeoutTimestamp = uint64(block.timestamp + 1 days) * 1e9;
      IbcUniversalPacketSender(mw).sendUniversalPacket(
          channelId, bytes32(destinationImplementation), message, timeoutTimestamp
      );
      return 0;
  }
  ```